torch/csrc/jit/codegen/cuda/scheduler/matmul.cpp

```diff
@@ -0,0 +1,222 @@
+#include <torch/csrc/jit/codegen/cuda/scheduler/matmul.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler/mma_utils.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler/utils.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+// Move the broadcast axes to the left on the specified number of inner
+// dimensions e.g.  (when number_of_inner_pos == 3):
+//      [... I0, B, I1] -> [... B, I0, I1]
+//  should probably be only used to order innermost mnk axes.
+void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
+  TORCH_INTERNAL_ASSERT(tv->nDims() >= number_of_inner_pos);
+  std::vector<int> broadcast_pos;
+  std::vector<int> nonbroadcast_pos;
+
+  for (auto i : c10::irange(number_of_inner_pos)) {
+    auto axis_idx = i - number_of_inner_pos;
+    auto id = tv->axis(axis_idx);
+    if (id->isBroadcast()) {
+      broadcast_pos.push_back(axis_idx);
+    } else {
+      nonbroadcast_pos.push_back(axis_idx);
+    }
+  }
+
+  auto combined_pos_vec = broadcast_pos;
+  combined_pos_vec.insert(
+      combined_pos_vec.end(), nonbroadcast_pos.begin(), nonbroadcast_pos.end());
+
+  std::unordered_map<int, int> order_map;
+  for (auto i : c10::irange(number_of_inner_pos)) {
+    order_map[combined_pos_vec.at(i)] = i - number_of_inner_pos;
+  }
+
+  // Apply ordering.
+  tv->reorder(order_map);
+}
+
+} // namespace
+
+void scheduleMatmul(
+    TensorView* c,
+    TensorView* a,
+    TensorView* b,
+    MmaBuilder& mma_builder,
+    MatMulTileOptions& gemm_tile) {
+  // Currently only support a, b, c as fusion inputs/outputs
+  //  aka. no prolog and epilog fusion yet.
+  TORCH_CHECK(
+      c->isFusionOutput() && a->isFusionInput() && b->isFusionInput(),
+      "not supporting matmul fusion yet");
+  TORCH_CHECK(c->definition() && c->definition()->isA<MmaOp>());
+
+  mma_builder.configureMma(c);
+
+  // Setup register and shared memory stages:
+  //   TODO: this section goes to a separate matmul util,
+  //   and needs more configurability.
+
+  // Setup accumulator register.
+  auto cc = c->cacheBefore();
+  mma_builder.accumulatorTv(cc);
+
+  // Get the input to the mma op.
+  auto mma = dynamic_cast<MmaOp*>(cc->definition());
+  TORCH_INTERNAL_ASSERT(mma != nullptr);
+  auto ab = mma->inA()->as<TensorView>();
+  auto bb = mma->inB()->as<TensorView>();
+
+  // Get exact configurations from mma builder.
+  auto mma_options = mma_builder.build();
```

Comment:
All the info written into mma_builder wouldn't be needed on the mma_op. 

`mma_builder` currently just serves as a `context` for all the swizzle interfaces and since now we also have schedule parameters class. I guess much of the information kept in mma_options could might as well just be moved to schedule params.

Reply:
Actually matmul_params is introduced in the next PR. I will just try to separate info into the params in follow ups if that's ok.